### ISSUE TITLE
Merging QLoRA weights to Qwen-2.5-VL-7B-Instruct does not change base model weights

### ISSUE THREAD

**Issue by harshit2997** [huggingface/peft#2586]
### System Info

Python 3.10.18

torch==2.6.0
peft==0.15.2
transformers==4.52.4
bitsandbytes==0.46.0

### Who can help?

Tagging @BenjaminBossan @sayakpaul for help here.

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [x] My own task or dataset (give details below)

### Reproduction

I have fine-tuned Qwen-2.5-VL-7B-Instruct on an image classification task using QLoRA. For this discussion, I will call the pre-trained 4-bit quantized model `model` which is obtained as below:

```
model_path = "Qwen/Qwen2.5-VL-7B-Instruct"
DEVICE = "cuda:0"

quantization_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_use_double_quant=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch.bfloat16,
)
processor = AutoProcessor.from_pretrained(model_path, trust_remote_code=True, min_pixels=1024*28*28, use_Fast=True)

model = Qwen2_5_VLForConditionalGeneration.from_pretrained(
    model_path, 
    device_map=DEVICE, 
    torch_dtype=torch.bfloat16,
    attn_implementation="flash_attention_2",
    trust_remote_code=True,
    quantization_config=quantization_config
).to(DEVICE)
```

After fine-tuning using SFTTrainer and PEFT (LoRA), I get the adapter weights stored at `adapter_path`. The relevant training code is:

```
peft_config = LoraConfig(
    lora_alpha=16,
    lora_dropout=0.05,
    r=8,
    bias="none",
    target_modules= 'all-linear',
    task_type="CAUSAL_LM",
)

peft_model = get_peft_model(model, peft_config)

training_args = SFTConfig(
  ## training params
)

trainer = SFTTrainer(
    model=model,
    args=training_args,
    train_dataset=messages_list_train,
    eval_dataset=messages_list_val,
    data_collator=collate_fn, ## a function I wrote to convert to tokens
    peft_config=peft_config,
    processing_class=processor.tokenizer,
)

trainer.train()
```

Once I get the adapter checkpoint at `adapter_path`, I load it using two different ways - both give the same metrics on my test set and perform better than the original pre-trained `model`. The two ways are:

```
## Method 1: use load_adapter
model.load_adapter(adapter_path)

## Method 2: use PeftModel class
model = PeftModel.from_pretrained(model, adapter_path)
```

However, when I try to merge the LoRA weights and get a standalone model, the resultant `merged_model` has the same weights as `model`:

```
merged_model = PeftModel.from_pretrained(model, adapter_path).merge_and_unload()

pt = sorted(list(model.state_dict().items()), key=lambda p:p[0])
mg = sorted(list(merged_model.state_dict().items()), key=lambda p:p[0])
## Below code runs without any assertion errors
for i in range(len(pt)):
    k1 = pt[i][0]
    k2 = mg[i][0]
    assert k1==k2
    assert (torch.abs(pt[i][1]-mg[i][1])).sum() == 0  ## Would have expected this line to raise assertion error
```

Why is this happening? Merging should have given the same results as the other two methods of loading LoRA weights. I also get this warning when merging but that seems to be expected according to [this issue](https://github.com/huggingface/peft/issues/2321).

`UserWarning: Merge lora module to 4-bit linear may get different generations due to rounding errors.`

### Expected behavior

Merged model should have different weights than the base model.

**Comment by harshit2997** [comment]
Even if I merge the adapter into the unquantized model (i.e. don't use `quantization_config=quantization_config` when creating `model`), the weights of `model` are same as `merged_model`.

**Comment by harshit2997** [comment]
My bad - I did not notice that `model` also keeps getting updated inplace hence I get the same weights in the end. `merged_model` gave the same metrics as the base model but that also seems to be a side effect of merging the adapter in the quantized model. When I merged it into the unquantized base model, I got even better metrics that the one I got when loading the adapter with the quantized base model without merging.

To anyone who reads this thread, my entire reason for merging was to be able to run the model in vLLM. vLLM does not support adding LoRA adapter to the image backbone. I didn't know this before fine tuning and I fine tuned all linear layers. So, the only way to use the trained adapter in vLLM was to merge it into the base model and create a new model. So, in the end, turns out that the only use of quantization for me was to have more efficient finetuning.
